website/src/documentation/programming-guide.md

```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
   }
 }
 ```  
+
+## 10. State and Timers {#state-and-timers}
+Beam's windowing and triggering facilities provide a powerful abstraction for grouping and aggregating unbounded input
+data based on timestamps. However there are aggregation use cases for which Beam's windows and triggers are not the best
+fit. Beam provides an API for manually managing per-key state, allowing for finer-grained control over aggregations.
+
+Beam's state API models state per key. To use the state API, you start out with a keyed `PCollection`, which in Java
+is modeled as a `PCollection<KV<K, V>>`. A `ParDo` processing this `PCollection` can now declare state variables. Inside
+the `ParDo` these state variables can be used to write or update state for the current key or to read previous state
+written for that key. State is always fully scoped only to the current processing key.
+
+Windowing can still be used together with stateful processing. All state for a key is scoped to the current window. This
+means that the first time a key is seen for a given window any state reads will return empty, and that a runner can
+garbage collect state when a window is completed.  Merging windows are not currently supported when using state and timers.
+
+
+### 10.1 Types of state {#types-of-state}
+Beam provides several types of state:
+
+#### ValueState
+A DoFn declares states to be accessed by creating final `StateSpec` member variables representing each state. Each
+state must be named using the `StateId` annotation; this name is unique to a ParDo in the graph and has no relation
+to other nodes in the graph. A `DoFn` can declare multiple state variables.
+
+If the type of the ValueState has a coder registered, then Beam will automatically infer the coder for the state value.
+For example, the following ParDo creates a single state variable that accumulates the number of elements seen.
```

Comment:
Worth explicitly calling out the lack of order? Although it doesnt matter for this use case, ValueState can be misused for other things when not aware of no order

Reply:
done